### PR TITLE
Block third party cookies

### PR DESCRIPTION
**Pull request by SlayterDev**
**Reviewer:** @kdzwinel 

## Description:

This PR adds blocking functionality for 3rd party cookies

## Steps to test this PR:
1. Visit https://privacy-test-pages.glitch.me/privacy-protections/storage-blocking
2. Click "Store data"
3. Click "Retrieve data"
4. Third party header and js cookie should fail

## Automated tests:
- [x] Unit tests
- [x] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [x] Get advice or leverage existing code
- [x] Agree on technical approach with reviewer (if the changes are nuanced)
- [x] Ensure that there is a testing strategy (and documented non-automated tests)
- [x] Ensure there is a documented monitoring strategy (if necessary)
- [x] Consider systems implications

### CODE REVIEW

File: shared/js/content-scripts/block-cookie.js
Comment (dharb) [inline shared/js/content-scripts/block-cookie.js:52]: Just a heads up that this is going to get blocked by sites with strict script-src CSPs in Firefox, take a look at what happens on twitter.com. Firefox offers a great workaround called [Xray vision](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Xray_vision). Check out how I used it in our GPC content script [here](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/develop/shared/js/content-scripts/GPC-moz.js) (as compared to the [content script](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/develop/shared/js/content-scripts/GPC.js) we use for chrome).

File: shared/js/background/chrome-events.es6.js
Comment (kdzwinel) [inline shared/js/background/chrome-events.es6.js:67]: nit: any reason why we are doing this check here and not after `if (cookieConfig.isExcluded(request.url)) return`? It seems like we can do it before we start looking for a header.

File: shared/js/background/utils.es6.js
Comment (kdzwinel) [inline shared/js/background/utils.es6.js:195]: hm, I wonder what will happen here for edge cases like IPs and chrome://newtab etc.

cc @sammacbeth

File: shared/js/background/storage/cookies.es6.js
Comment (kdzwinel) [inline shared/js/background/storage/cookies.es6.js:58]: nit: we are always adding to that array but never clear it - does it mean that the array increases in size every time list is fetched?

File: shared/js/background/storage/cookies.es6.js
Comment (kdzwinel) [inline shared/js/background/storage/cookies.es6.js:64]: this assumes that excluded domains will always be top level and I wonder if that's correct 🤔 maybe in some cases we only want to safelist `accounts.google.com` but not the whole `google.com`?

File: shared/js/content-scripts/block-cookie.js
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:5]:
```suggestion
        let parentFrame = (document !== window.top) && window.parent
```

Is this file covered by our linting?

File: shared/js/content-scripts/block-cookie.js
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:27]: can we move this code in huge strings to a function like:

```js
function injectedClearCookies() {
   // Clear previously set cookies
  …
}

…

const scriptString = `(function() {${injectedClearCookies.toString()}()})()`

```

This way syntax will be linted and higlighted in diffs 🌈

File: shared/js/content-scripts/block-cookie.js
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:38]: I haven't realized it's that complicated to clear cookies. I thought it will be `document.cookie = ''` :)

BTW I remembered about https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Clear-Site-Data - we could inject this header to force browser to clear cookeis - not sure if that'd make anything easier, but just FYI

File: shared/js/content-scripts/block-cookie.js
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:55]: ditto re string -> function

File: shared/js/content-scripts/block-cookie.js
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:58]: what will happen to session cookies?

**Review by kdzwinel — COMMENTED** [review]
Lookin' good 💪 I left couple of recommendations.

File: shared/js/content-scripts/block-cookie.js
Comment (SlayterDev) [reply shared/js/content-scripts/block-cookie.js]: This is magical ✨

File: shared/js/background/chrome-events.es6.js
Comment (sammacbeth) [inline shared/js/background/chrome-events.es6.js:232]: Do we need to get the `frameUrl` in the content script? According to [MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/MessageSender) the `sender.url` will return the frame URL if it is sent from a frame:

> If the sender is a script running in a web page (including content scripts as well as normal page scripts), then url will be the web page URL. If the script is running in an iframe, url will be the iframe's URL.

File: shared/js/content-scripts/block-cookie.js
Comment (sammacbeth) [inline shared/js/content-scripts/block-cookie.js:76]: See earlier comment in the `onMessage` listener - the frameURL should automatically be set on the `sender`.

File: shared/js/content-scripts/block-cookie.js
Comment (sammacbeth) [inline shared/js/content-scripts/block-cookie.js:31]: If I understand this correctly, you are unsetting cookies for all possible subdomains of `window.location.hostname`. Why is that necessary? Can we not just use the existing domain field?

File: shared/js/content-scripts/block-cookie.js
Comment (sammacbeth) [inline shared/js/content-scripts/block-cookie.js:60]: Will this convert session cookies to normal cookies?

File: shared/js/content-scripts/block-cookie.js
Comment (sammacbeth) [inline shared/js/content-scripts/block-cookie.js:46]: How about using `;max-age=604800`? I believe the browser will ignore 'expiries' if 'max-age' is specified ([Source](https://mrcoles.com/blog/cookies-max-age-vs-expires/))

**Review by sammacbeth — COMMENTED** [review]
Looks good! I left a few comments.

File: shared/js/background/chrome-events.es6.js
Comment (SlayterDev) [reply shared/js/background/chrome-events.es6.js:65]: I went back and forth on this a bit but think you may be right here. cc @kdzwinel

File: shared/js/content-scripts/block-cookie.js
Comment (SlayterDev) [reply shared/js/content-scripts/block-cookie.js:31]: Initially it was just being over cautious but based on the rest of the implementation the current domain should suffice.

File: shared/js/background/utils.es6.js
Comment (SlayterDev) [reply shared/js/background/utils.es6.js:195]: Do we need to worry about IP addresses in this implementation? If so is there an easy check if a domain is an IP address so we can compare? Also cc @Charlie-belmer who originally wrote this method for UA rotation.

File: shared/js/background/storage/cookies.es6.js
Comment (kdzwinel) [inline shared/js/background/storage/cookies.es6.js:64]: Ah, now we are not giving ourselves an option to exclude whole domain with subdomains (we will have to all full hosts). Maybe that's OK, but maybe we should do incremental check instead here? e.g. "sub.domain.com" -> check("sub.domain.com") + check("domain.com")

File: shared/js/background/utils.es6.js
Comment (kdzwinel) [reply shared/js/background/utils.es6.js:195]: IMO we should be taking IPs into account, tldts returns `isIP` property so that should be straightforward.

As for the private domains, I think we should stick to the standard definition of a third party, as used by other browsers, and don't use private domains for now. It's a good idea for a follow up though.

File: shared/js/content-scripts/block-cookie.js
Comment (kdzwinel) [inline shared/js/content-scripts/block-cookie.js:68]: nit: we should remove console.logs in content scripts before merging

**Review by kdzwinel — COMMENTED** [review]
I took another pass here and found some issues:

1. We are only blocking cookies from being set, but we should also block them from being sent. Right now, if someone goes to e.g. google.com, we will allow them to set cookies (first party rule), but we will then also allow those cookies to be sent on other sites!

2. On https://privacy-test-pages.glitch.me/privacy-protections/storage-blocking/ it looks like JS cookie in a 3p iframe is being set - it shouldn't. This is possibly related to an error that I'm getting in the console - " d.shift is not a function".

3. Tests: we should add an integration test for https://privacy-test-pages.glitch.me/privacy-protections/storage-blocking/ and, ideally, unit tests for manipulating cookies as this code looks a bit scary 😬

**Comment by sammacbeth** [comment]
@kdzwinel re: tests, I've opened #544 which does integration tests against privacy-test-pages for cookie blocking. This is intended to cover the tests needed for my project, but it also easily covers 3rd party cookie blocking, so I added those tests too. We could cherry-pick that commit into the PR to add tests.

File: shared/js/background/chrome-events.es6.js
Comment (sammacbeth) [inline shared/js/background/chrome-events.es6.js:79]: Firefox doesn't support the `extraHeaders` option here which causes an exception when the extension is started (`Uncaught Error: Type error for parameter extraInfoSpec (Error processing 2: Invalid enumeration value "extraHeaders") for webRequest.onHeadersReceived.`)

You can feature detect support for these flags via `chrome.webRequest.OnHeadersReceivedOptions`. e.g.
```js
const extraInfoSpec = ['blocking', 'responseHeaders'];
if (chrome.webRequest.OnHeadersReceivedOptions.EXTRA_HEADERS) {
   extraInfoSpec.push(chrome.webRequest.OnHeadersReceivedOptions.EXTRA_HEADERS)
}
```

File: shared/js/content-scripts/block-cookie.js
Comment (sammacbeth) [inline shared/js/content-scripts/block-cookie.js:24]:
```suggestion
```

(Looks like this should have been removed in refactor)

File: browsers/chrome/manifest.json
Comment (sammacbeth) [inline browsers/chrome/manifest.json:43]: Should this also be included in the Firefox manifest?

File: shared/js/content-scripts/block-cookie.js
Comment (sammacbeth) [inline shared/js/content-scripts/block-cookie.js:28]:
```suggestion
        document.__defineSetter__('cookie', function (value) { })
        document.__defineGetter__('cookie', () => '')
```

When testing this code I saw that `document.cookie` somehow became `undefined`. This broke scripts in the iframe because, according to the spec `document.cookie` always returns a string. Adding a stubbed getter seems to fix this.

File: browsers/chrome/manifest.json
Comment (SlayterDev) [reply browsers/chrome/manifest.json:43]: Yes 🤦‍♂️

File: shared/js/content-scripts/block-cookie.js
Comment (sammacbeth) [inline shared/js/content-scripts/block-cookie.js:43]: This doesn't look right (comparing a `Date` with the number of seconds in a week).

Comment (SlayterDev) [reply shared/js/content-scripts/block-cookie.js:43]: Yes still working on this as part of the session cookie fix 👍

**Comment by SlayterDev** [comment]
I've commited a change to use the TR block list for cookie blocking. Feel free to review it but it is currently untested due to intermittent power issues here.

File: shared/js/background/chrome-events.es6.js
Comment (kdzwinel) [inline shared/js/background/chrome-events.es6.js:81]: I found out that this will only remove singe `set-cookie` header while in the wild there can be multiple:

3-set-cookie

We should probably `.filter()` all of those out here.

File: shared/js/background/chrome-events.es6.js
Comment (sammacbeth) [inline shared/js/background/chrome-events.es6.js:99]:
```suggestion
        return { responseHeaders }
```

File: shared/js/background/chrome-events.es6.js
Comment (sammacbeth) [inline shared/js/background/chrome-events.es6.js:72]: Is this left here by mistake?

File: shared/js/background/chrome-events.es6.js
Comment (sammacbeth) [inline shared/js/background/chrome-events.es6.js:87]: As this block comes up 3 times in this file, could we extract this into a function that just returns if we're in the cookie blocking experiment?

File: shared/js/background/chrome-events.es6.js
Comment (sammacbeth) [reply shared/js/background/chrome-events.es6.js:99]: I would also consider only attaching `responseHeaders` to the `BlockingResponse` if we actually change the headers. This will prevent any potential overhead for unchanged requests.

File: shared/js/background/chrome-events.es6.js
Comment (sammacbeth) [inline shared/js/background/chrome-events.es6.js:267]:
```suggestion
                action.isThirdParty = !utils.isFirstParty(sender.url, sender.tab.url))
```

File: shared/js/background/chrome-events.es6.js
Comment (sammacbeth) [inline shared/js/background/chrome-events.es6.js:270]:
```suggestion
                action.shouldBlock = !cookieConfig.isExcluded(sender.url) && trackerutils.isTracker(sender.url))
```

File: shared/js/content-scripts/block-cookie.js
Comment (sammacbeth) [inline shared/js/content-scripts/block-cookie.js:24]: It occurs to me that this may cause issues for sites that the user also visits in first party context. Consider a site that embeds a 3rd party iframe from YouTube. This code will wipe out all cookies except those that are http-only. On a subsequent visit to Youtube, there may be some breakage caused by the absense of these cookies.

Perhaps it is sufficient just to block the cookie setter and getter in this case?

Comment (SlayterDev) [reply shared/js/content-scripts/block-cookie.js:24]: Hmm this is a good point. What are your thoughts @kdzwinel?

Comment (kdzwinel) [reply shared/js/content-scripts/block-cookie.js:24]: Hm, good point. Those cookies will be still saved in a browser but should not be shared in the third party context. My only worry is that, as far as I remember, the removal of all 3p cookies was helping us to avoid the timing issue (i.e. if we injected our script after some cookies were already set it's OK because we would remove them). Am I recalling this correctly? Is this still the case?

Comment (SlayterDev) [reply shared/js/content-scripts/block-cookie.js:24]: Yes this is still here to take care of any race conditions

Comment (sammacbeth) [reply shared/js/content-scripts/block-cookie.js:24]: I still think this will cause breakage. For example, on YouTube there is a consent cookie set as not http-only, which would therefore be wiped by this. This means that everytime you visit a page with embedded youtube iframe, your consent settings would get wiped.

I think fixes for race conditions are planned, so we can wait for a proper solution there before addressing this.

Comment (kdzwinel) [reply shared/js/content-scripts/block-cookie.js:24]: Hm, OK lets drop removing cookies for now. We have to note this gap in the relevant task and update tech design.

File: integration-test/background/storage.js
Comment (sammacbeth) [reply integration-test/background/storage.js:39]: These are the original tests I wrote in #544 against the old tech designs - these will probably need updating for the updated policy.

File: shared/js/background/storage/cookies.es6.js
Comment (sammacbeth) [inline shared/js/background/storage/cookies.es6.js:76]: This function is O(N*M) in the worst case, where N is the length of `excludedDomains` and M is the subdomain depth of the given domain. As this is run for every third-party request, then most calls will run to the worst case complexity, and this function will be called often. 

I think it would make sense to optimise this a bit. Using a trie you should be able to get O(log M) performance on average.

File: shared/js/background/chrome-events.es6.js
Comment (sammacbeth) [inline shared/js/background/chrome-events.es6.js:103]:
```suggestion
            if (!cookieConfig.isExcluded(request.url) && trackerutils.isTracker(request.url)) {
                responseHeaders = responseHeaders.filter(header => header.name.toLowerCase() !== 'set-cookie')
                return { responseHeaders }
            }
            ```

This avoids returning if response headers are unchanged.

**Review by sammacbeth — CHANGES_REQUESTED** [review]

File: shared/js/background/storage/cookies.es6.js
Comment (SlayterDev) [reply shared/js/background/storage/cookies.es6.js:76]: I'm having issues imagining the trie implementation (my theory is rather rusty). Would you be able to go a bit more in depth. Do you think we'll be hitting the worst case most often because most cookies will be excluded by the root domain (i.e. example.com vs test.example.com)? If so the other idea I have is recursing to the root domain first and bubbling up from there. However I think some granularity in the exclusions could be lost this way?

Comment (sammacbeth) [reply shared/js/background/storage/cookies.es6.js:76]: I would organise the exclude list by reversed domain parts:
```json
{
  "com": {
     "google": {
        "accounts": true
     },
     "googleapis": true
}
```

Then you get worst-case O(M), but constant time average case with a lookup algorithm like:
```js
const parts = domain.split('.').reverse()
let ptr = this.excludedDomainsLookup;
for (const i = 0; i < parts.length; i++) {
  ptr = ptr[parts[i]]
  if (ptr === true) { // matching subdomain
     return true
  } else if (ptr === undefined) { // empty leaf
     return false;
  }
}
```
(warning: untested code that I just dropped in here, might not work at all!)

**Comment by sammacbeth** [comment]
Superseded by #570

**Pull request closed without merging**
